RyuSocks.Test/Auth/UsernameAndPasswordTests.cs

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) RyuSOCKS
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <https://www.gnu.org/licenses/>.
+ */
+
+using RyuSocks.Auth;
+using RyuSocks.Packets.Auth.UsernameAndPassword;
+using RyuSocks.Test.Utils;
+using System;
+using System.Collections.Generic;
+using System.Security.Authentication;
+using System.Text;
+using Xunit;
+
+namespace RyuSocks.Test.Auth
+{
+    public class UsernameAndPasswordTests
+    {
+        [Theory]
+        [InlineData(new byte[] { 0xFF, 0xFF, 0xAA, 0x00, 0xCC }, false)]
+        [InlineData(new byte[] { 0x01, 0x00, 0x00, 0x00, 0x00 }, true)]
+        public void Authenticate_RandomPackages(byte[] incomingPacket, bool isValidInput)
+        {
+            UsernameAndPassword usernameAndPassword = new()
+            {
+                Database = [],
+                IsClient = false
+            };
+            if (!isValidInput)
+            {
+                Assert.Throws<InvalidOperationException>(() => usernameAndPassword.Authenticate(incomingPacket, out _));
+            }
+            else
+            {
+                Assert.Throws<AuthenticationException>(() => usernameAndPassword.Authenticate(incomingPacket, out _));
+            }
+        }
+
+        [Theory]
+        [UsernameAndPasswordRandomUsernameAndPasswords(20)]
+        public void Authenticate_FullExchange(string username, string password)
```

Comment:
Same as before, make sure the method name states what you expect your method to do here.